### PR TITLE
Core: check that project schema and selected columns cannot be specified at the same time

### LINKED ISSUE #1304: `select` will overwrite `project` whatever the order in TableScan
`select` will overwrite `project` whatever the order.
-> table.newScan().select("id").project(SCHEMA.select("data"))
Expected: "data"
Actual: "id"

Should we fix this? Invalid `select` when `project`?

### PR DESCRIPTION
**Pull request by JingsongLi**
Fixes #1304

### CODE REVIEW

File: core/src/main/java/org/apache/iceberg/BaseTableScan.java [hunk 1]
```diff
@@ -159,6 +159,10 @@ public TableScan option(String property, String value) {
 
   @Override
   public TableScan project(Schema projectedSchema) {
+    if (context.selectedColumns() != null) {
+      throw new IllegalStateException("Cannot project schema when selected columns is specified.");
+    }
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/BaseTableScan.java:164]: We would normally use a `Precondition` check here:

```java
Preconditions.checkState(context.selectedColumns() == null, "Cannot ...");
```

I was also going to say that I think we should check the reverse as well, in case `project` is called before `select`, but it looks like the projection is tracked outside of the context for some reason. I don't think that's quite right. We probably want to move `project` to the context as well.

@edgarRd, do you know why `project` doesn't use the scan context?

Comment (JingsongLi) [reply core/src/main/java/org/apache/iceberg/BaseTableScan.java:164]: Another thing is: `BaseTableScan.lazyColumnProjection` looks weird. It looks only work for `DataTableScan`. Because other table scan like `MetadataTableScan`, the file/expected scan schema is different from `table.schema()`. But in `BaseTableScan.lazyColumnProjection`, it use `table.schema()` to select. And for `rowFilter`, it looks only works for `DataTableScan` selection, should not affect `MetadataTableScan`s selection.

In other words, we may need to implement different selections for various scans.

Comment (edgarRd) [reply core/src/main/java/org/apache/iceberg/BaseTableScan.java:164]: This was the idea on not putting schema in the context: https://github.com/apache/iceberg/pull/1115#discussion_r441068118 but I guess it could be part of the context, it would need to be initialized and in this case another context would need to be created with the new schema from the projection.

Comment (rdblue) [reply core/src/main/java/org/apache/iceberg/BaseTableScan.java:164]: Thanks @edgarRd! I think it makes sense to move that into the context and initialize it as null. If the projection is null, then use the column list. If not, then use the projection. And then both methods can validate that the other has not been called.

I think it would be okay to do this even if it only works for `DataTableScan`. If the other scans don't support it, then they should throw exceptions instead of allowing the projection to be set.

Comment (edgarRd) [reply core/src/main/java/org/apache/iceberg/BaseTableScan.java:164]: Yeah, that makes sense @rdblue. Should we have another PR for that change?

Comment (JingsongLi) [reply core/src/main/java/org/apache/iceberg/BaseTableScan.java:164]: Hi @edgarRd , If you don't mind, I think I can finish it in this PR.

File: core/src/main/java/org/apache/iceberg/TableScanContext.java [hunk 2]
```diff
@@ -107,16 +112,27 @@ boolean returnColumnStats() {
...
-        caseSensitive, returnColumnStats, selectedColumns, options, fromSnapshotId, toSnapshotId);
+        caseSensitive, returnColumnStats, projectedSchema, selectedColumns, options, fromSnapshotId, toSnapshotId);
   }
 
   Collection<String> selectedColumns() {
     return selectedColumns;
   }
 
   TableScanContext selectColumns(Collection<String> columns) {
+    Preconditions.checkState(projectedSchema == null, "Cannot selected columns when project schema is specified.");
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/TableScanContext.java:123]: Nit: should be `Cannot select columns`, and you can omit the ending `.`

File: core/src/main/java/org/apache/iceberg/TableScanContext.java [hunk 3]
```diff
@@ -107,16 +112,27 @@ boolean returnColumnStats() {
...
-        caseSensitive, colStats, columns, options, fromSnapshotId, toSnapshotId);
+        caseSensitive, colStats, null, columns, options, fromSnapshotId, toSnapshotId);
+  }
+
+  Schema projectedSchema() {
+    return projectedSchema;
+  }
+
+  TableScanContext project(Schema schema) {
+    Preconditions.checkState(selectedColumns == null, "Cannot project schema when selected columns is specified.");
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/TableScanContext.java:133]: How about `Cannot set projection schema when columns are selected`

File: core/src/main/java/org/apache/iceberg/BaseTableScan.java [hunk 4]
```diff
@@ -284,6 +284,8 @@ public String toString() {
   private Schema lazyColumnProjection() {
     Collection<String> selectedColumns = context.selectedColumns();
     if (selectedColumns != null) {
+      // select columns only works for data table scans, so it is OK to select on table.schema().
+
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/BaseTableScan.java:288]: I don't understand why select only works for data table scans. I updated this method to use the `schema` that is passed into the constructor and all Iceberg tests pass. By updating this to use the schema passed in instead of the table schema, we can support either project or select, depending on what the caller chooses.

Then we can remove the overrides that throw `UnsupportedOperationException`.

File: core/src/main/java/org/apache/iceberg/BaseTableScan.java [hunk 5]
```diff
@@ -301,6 +303,9 @@ private Schema lazyColumnProjection() {
       requiredFieldIds.addAll(selectedIds);
 
       return TypeUtil.select(table.schema(), requiredFieldIds);
+    } else if (context.projectedSchema() != null) {
+
+      return context.projectedSchema();
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/BaseTableScan.java:308]: Nit: I find this more readable if the blank line is before the `else` to distinguish between the previous block and the else case.

**Comment by rdblue** [comment]
Thanks @JingsongLi. I think this is close, but I don't think that we need to restrict `select` to only data table scans. We should be able to use it when planning any scan by updating schema pruning to use the schema passed in to `BaseTableScan` when it is created.

**Comment by JingsongLi** [comment]
Thanks @rdblue for your review, yes, we can support `select` better, should be able to use it for all scans. I'll try to do it.

File: core/src/main/java/org/apache/iceberg/BaseTableScan.java (hunk 4 above)
Comment (JingsongLi) [reply core/src/main/java/org/apache/iceberg/BaseTableScan.java:288]: The problem is on `rowFilter`.
IIUC, `rowFilter` is based on table data for every scans, so:
- if it is data table scan, we should add the field ids of `rowFilter` to `requiredFieldIds`.
- if it is not data table scan, we should not, because the schema not contains fields of `rowFilter`.

File: core/src/main/java/org/apache/iceberg/BaseTableScan.java [hunk 6]
```diff
@@ -106,6 +106,14 @@ protected abstract TableScan newRefinedScan(
       TableOperations ops, Snapshot snapshot, Expression rowFilter,
       boolean ignoreResiduals, boolean caseSensitive, boolean colStats);
 
+  /**
+   * The {@link #filter(Expression)} is based on the table records. So the data table scan need select all of the
+   * filter columns, other scans don't need to.
+   */
+  protected boolean isDataTableScan() {
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/BaseTableScan.java:113]: I'm not sure I understand why this is added. Shouldn't the filter (which is unbound) be bound to the scan's schema and not the underlying table's schema?

File: core/src/main/java/org/apache/iceberg/BaseTableScan.java [hunk 7]
```diff
@@ -286,21 +294,26 @@ private Schema lazyColumnProjection() {
     if (selectedColumns != null) {
       Set<Integer> requiredFieldIds = Sets.newHashSet();
 
-      // all of the filter columns are required
-      requiredFieldIds.addAll(
-          Binder.boundReferences(table.schema().asStruct(),
-              Collections.singletonList(context.rowFilter()), context.caseSensitive()));
+      if (isDataTableScan()) {
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/BaseTableScan.java:297]: Why not do this for all table scans?

**Comment by rdblue** [comment]
Looks good, except for the new method. Could you help me understand your rationale for only getting filter columns when performing a data table scan?

**Comment by JingsongLi** [comment]
Sorry, It seems that I misunderstood it.
Is the filter applied to the schema of scan or the schema of table?
For example, a table schema: {id -> int, data -> string}, but the schema of a `FilesTableScan` is {content -> int, file_path -> string, file_size -> long, ....}. (Defined in `DataFile`)

So IIUC, filters on `FilesTableScan` should be `file_path=...` or `file_size>.. and file_size<...`. I got it wrong because there was something wrong with the sample in the test class: `TestMetadataTableScans`, in this class, all filters on id...

I'll change the code and fix the tests.

**Comment by rdblue** [comment]
Thanks, @JingsongLi! I can see why that test was misleading. Thanks for fixing this, I'll merge.
